### PR TITLE
Xai

### PR DESCRIPTION
**Pull request by rvandewater**

### CODE REVIEW

File: configs/tasks/common/CrossValidation.gin
Comment (rvandewater) [inline configs/tasks/common/CrossValidation.gin:1]: should be reset to 5

File: experiments/benchmark_classification.yml
Comment (rvandewater) [inline experiments/benchmark_classification.yml:1]: reset it

File: icu_benchmarks/data/preprocessor.py
Comment (rvandewater) [inline icu_benchmarks/data/preprocessor.py:1]: why like this

File: icu_benchmarks/data/loader.py [hunk 1]
```diff
@@ -160,6 +169,138 @@ def to_tensor(self):
             return from_numpy(data), from_numpy(labels)
 
 
+"""
```
Comment (rvandewater) [inline icu_benchmarks/data/loader.py:172]: why commented out?

File: icu_benchmarks/data/Checking data-Copy1.ipynb
Comment (rvandewater) [inline icu_benchmarks/data/Checking data-Copy1.ipynb:1]: is this necessary?

File: icu_benchmarks/cross_validation.py
Comment (rvandewater) [inline icu_benchmarks/cross_validation.py:1]: change back

File: icu_benchmarks/models/constants.py [hunk 2]
```diff
@@ -64,11 +66,11 @@ class DLMetrics:
 
     BINARY_CLASSIFICATION_TORCHMETRICS = {
         "AUC": AUROC(task="binary"),
-        "PR": TorchMetricsAveragePrecision(task="binary"),
-        "PrecisionRecallCurve": TorchMetricsPrecisionRecallCurve(task="binary"),
+        # "PR": TorchMetricsAveragePrecision(task="binary"),
```
Comment (rvandewater) [inline icu_benchmarks/models/constants.py:69]: why is this commented out?

File: configs/tasks/Regression.gin
Comment (rvandewater) [inline configs/tasks/Regression.gin:1]: change back loss for dl prediction wrapper

File: icu_benchmarks/models/constants.py [hunk 3]
```diff
@@ -14,8 +19,6 @@
 )
 from torchmetrics.classification import (
     AUROC,
-    AveragePrecision as TorchMetricsAveragePrecision,
-    PrecisionRecallCurve as TorchMetricsPrecisionRecallCurve,
     CalibrationError,
```
Comment (rvandewater) [inline icu_benchmarks/models/constants.py:22]: Why did you remove those metrics?

File: icu_benchmarks/models/dl_models.py [hunk 4]
```diff
@@ -280,3 +318,384 @@ def forward(self, x):
         o = o.permute(0, 2, 1)  # Permute to channel last
         pred = self.logit(o)
         return pred
+
+
+'''
+@gin.configurable
```
Comment (rvandewater) [inline icu_benchmarks/models/dl_models.py:324]: why is it commented out?

File: icu_benchmarks/models/wrappers.py [hunk 5]
```diff
@@ -356,14 +385,1007 @@ def step_fn(self, element, step_prefix=""):
...
+            marker="o",
+            color="skyblue",
+            linestyle="-",
+            linewidth=2,
+            markersize=8,
+        )
+        plt.xlabel("Feature")
+        plt.ylabel("{} Attribution".format(method_name))
+        plt.title("{} Attribution Values".format(method_name))
+        plt.xticks(
```
Comment (rvandewater) [inline icu_benchmarks/models/wrappers.py:553]: plotting should not be done in here

File: icu_benchmarks/run.py [hunk 6]
```diff
@@ -73,7 +82,7 @@ def main(my_args=tuple(sys.argv[1:])):
             else "None"
         }
     )
-
+    random_model_dir = args.random_model
```
Comment (rvandewater) [inline icu_benchmarks/run.py:85]: what is this?

File: icu_benchmarks/run_utils.py [hunk 7]
```diff
@@ -25,33 +27,134 @@ def build_parser() -> ArgumentParser:
...
+        default=False,
+        action=BOA,
+        help="Set to log verbosly. Disable for clean logs.",
+    )
     parser.add_argument("--cpu", default=False, action=BOA, help="Set to use CPU.")
     parser.add_argument("-db", "--debug", default=False, action=BOA, help="Set to load less data.")
     parser.add_argument("--reproducible", default=True, action=BOA, help="Make torch reproducible.")
-    parser.add_argument("-lc", "--load_cache", default=False, action=BOA, help="Set to load generated data cache.")
-    parser.add_argument("-gc", "--generate_cache", default=False, action=BOA, help="Set to generate data cache.")
+    parser.add_argument(
```
Comment (rvandewater) [inline icu_benchmarks/run_utils.py:72]: this seems kind of messy, are you sure the autoformatting worked here? Perhaps you can change it back

File: icu_benchmarks/run_utils.py [hunk 8]
```diff
@@ -157,6 +272,85 @@ def aggregate_results(log_dir: Path, execution_time: timedelta = None):
     wandb_log(json.loads(json.dumps(accumulated_metrics, cls=JsonResultLoggingEncoder)))
 
 
+def plot_XAI_Metrics(accumulated_metrics, log_dir_plots):
```
Comment (rvandewater) [inline icu_benchmarks/run_utils.py:275]: should not be in here

File: icu_benchmarks/tuning/hyperparameters.py [hunk 9]
```diff
@@ -36,6 +36,8 @@ def choose_and_bind_hyperparameters(
     debug: bool = False,
     verbose: bool = False,
     wandb: bool = False,
+    pytorch_forecasting: bool = False,
```
Comment (rvandewater) [inline icu_benchmarks/tuning/hyperparameters.py:39]: I am not sure about putting this here as it makes the code quite inflexible. Are you able to use isinstance for example?

File: icu_benchmarks/models/wrappers.py [hunk 10]
```diff
@@ -356,14 +385,1007 @@ def step_fn(self, element, step_prefix=""):
...
+
+        pred_deltas_cpu = torch.tensor(pred_deltas).cpu().numpy()
+        att_sums_cpu = torch.tensor(att_sums).cpu().numpy()
+
+        similarities.append(similarity_func(pred_deltas_cpu, att_sums_cpu))
+
+        score = np.nanmean(similarities)
+        return score
+
+    def Data_Randomization(
```
Comment (rvandewater) [inline icu_benchmarks/models/wrappers.py:1113]: put in utils

File: icu_benchmarks/models/wrappers.py [hunk 11]
```diff
@@ -356,14 +385,1007 @@ def step_fn(self, element, step_prefix=""):
...
+            attribution = (attribution - min_val) / (max_val - min_val)
+            random_attr = random_attr.flatten()
+            min_val = np.min(random_attr)
+            max_val = np.max(random_attr)
+            random_attr = (random_attr - min_val) / (max_val - min_val)
+
+            score = similarity_func(random_attr, attribution)
+        return score
+
+    def Relative_Stability(
```
Comment (rvandewater) [inline icu_benchmarks/models/wrappers.py:1196]: put in utils

File: icu_benchmarks/models/wrappers.py [hunk 12]
```diff
@@ -356,14 +385,1007 @@ def step_fn(self, element, step_prefix=""):
...
+        self,
+        x,
+        attribution,
+        explain_method,
+        method_name,
+        dataloader=None,
+        thershold=0.5,
```
Comment (rvandewater) [inline icu_benchmarks/models/wrappers.py:1203]: typo

File: icu_benchmarks/models/wrappers.py [hunk 13]
```diff
@@ -356,14 +385,1007 @@ def step_fn(self, element, step_prefix=""):
...
+                x_original[close_indices, :, :].detach(),
+                x_preturb[close_indices, :, :].detach(),
+                attribution,
+                att_preturb,
+                input=True,
+            )
+
+            ROS = relative_stability_objective(
+                y_pred[close_indices],
+                y_pred_preturb[close_indices],
```
Comment (rvandewater) [inline icu_benchmarks/models/wrappers.py:1317]: typo? Should be perturb I think

File: configs/prediction_models/common/DLTuning.gin
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply configs/prediction_models/common/DLTuning.gin:1]: changed to 30

File: configs/tasks/Regression.gin
Comment (youssefmecky96) [reply configs/tasks/Regression.gin:1]: reverted loss to mse loss again

File: configs/tasks/common/CrossValidation.gin
Comment (youssefmecky96) [reply configs/tasks/common/CrossValidation.gin:1]: reverted

File: experiments/benchmark_classification.yml
Comment (youssefmecky96) [reply experiments/benchmark_classification.yml:1]: done

File: experiments/benchmark_regression.yml
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply experiments/benchmark_regression.yml:1]: done

File: icu_benchmarks/cross_validation.py
Comment (youssefmecky96) [reply icu_benchmarks/cross_validation.py:1]: changed

File: icu_benchmarks/data/Checking data-Copy1.ipynb
Comment (youssefmecky96) [reply icu_benchmarks/data/Checking data-Copy1.ipynb:1]: removed it

File: icu_benchmarks/data/constants.py [hunk 14]
```diff
@@ -15,3 +15,7 @@ class VarType:
     group = "GROUP"
     sequence = "SEQUENCE"
     label = "LABEL"
+
+
+class FeatType:
```
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply icu_benchmarks/data/constants.py:20]: This is from when I had the TFT built from scratch so can remove it

File: icu_benchmarks/data/loader.py (hunk 1 above)
Comment (youssefmecky96) [reply icu_benchmarks/data/loader.py:172]: if you are referring to  # [self.vars["GROUP"]] this is from the YAIB main branch I didn't add the comment

File: icu_benchmarks/data/preprocessor.py
Comment (youssefmecky96) [reply icu_benchmarks/data/preprocessor.py:1]: This was from the time we had the discussion about adding the min max dynamically instead of declaring  it in the config file

File: icu_benchmarks/models/constants.py (hunk 3 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/constants.py:22]: I had them commented out as when I was predicting sepsis and tried to use them they had errors and then with the formatting it showed as not used I uncommented them now and reimported them

File: icu_benchmarks/models/constants.py (hunk 2 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/constants.py:69]: Mentioned above

File: icu_benchmarks/models/similarity_func.py
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply icu_benchmarks/models/similarity_func.py:1]: Added quantum to environment.yml and imported directly from there

File: icu_benchmarks/run.py (hunk 6 above)
Comment (youssefmecky96) [reply icu_benchmarks/run.py:85]: This is the path of the model trained on random labels which is needed when calculating the Data randomization distance

File: icu_benchmarks/run_utils.py (hunk 7 above)
Comment (youssefmecky96) [reply icu_benchmarks/run_utils.py:72]: changed it manually

File: icu_benchmarks/run_utils.py (hunk 8 above)
Comment (youssefmecky96) [reply icu_benchmarks/run_utils.py:275]: moved it to scripts/plotting although this is an extra I didn't really use this plotting function

File: icu_benchmarks/tuning/hyperparameters.py (hunk 9 above)
Comment (youssefmecky96) [reply icu_benchmarks/tuning/hyperparameters.py:39]: The thing is train_common doesn't take the model class in the beginning just has the name and then creates a model so is instance wouldn't work but here it gets this from the run command so this doesn't come from each function it is defined by the person running the code once in the beginning and then is pushed through everything

File: icu_benchmarks/models/custom_metrics.py [hunk 15]
```diff
@@ -130,3 +136,423 @@ def feature_helper(self, trainer, step_prefix):
         else:
             feature_names = trainer.test_dataloaders.dataset.features
         return feature_names
+
+
+# XAI Metrics
+
+
+class Faithfulness(EpochMetric):
```
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply icu_benchmarks/models/custom_metrics.py:144]: These ones can't be directly imported as I needed to modify them to get them to work with YAIB and pytorch forecasting , they are also the same ones in the wrapper.py however I added them here as an extra option to calculate the XAI metrics as a metric instead of manually. I will add the citations

File: icu_benchmarks/models/dl_models.py (hunk 4 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/dl_models.py:324]: This is the model built from scratch in the beginning I kept it as I thought this could be handy at some point however only used the model from pytorch forecasting in the end tell me if you think it is better to remove it

File: icu_benchmarks/models/wrappers.py (hunk 5 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/wrappers.py:553]: this was a check for me can remove the plots

File: icu_benchmarks/models/wrappers.py (hunk 12 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/wrappers.py:1203]: corrected to threshold

File: icu_benchmarks/models/wrappers.py (hunk 10 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/wrappers.py:1113]: moved to utils

File: icu_benchmarks/models/wrappers.py (hunk 13 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/wrappers.py:1317]: fixed

File: experiments/demo_benchmark_classification.yml [hunk 16]
```diff
@@ -11,15 +11,15 @@ command:
...
+  - -lc 
 method: grid
 name: yaib_demo_classification_benchmark
 parameters:
   data_dir:
     values:
       - demo_data/mortality24/eicu_demo
       - demo_data/mortality24/mimic_demo
-      - demo_data/aki/eicu_demo
+     #fails for some reason - demo_data/aki/eicu_demo
```
Comment (rvandewater) [inline experiments/demo_benchmark_classification.yml:22]: Would warrant further investigation

File: environment.yml [hunk 17]
```diff
@@ -13,7 +13,7 @@ dependencies:
   - gin-config=0.5.0
   - ignite=0.4.11
   - pytorch=2.0.1
-  - pytorch-cuda=11.8
+ # - pytorch-cuda=11.8
```
Comment (rvandewater) [inline environment.yml:16]: I think this is because the other packages already install a certain cuda?

File: experiments/demo_benchmark_regression.yml [hunk 18]
```diff
@@ -19,16 +19,16 @@ parameters:
...
+   #   - demo_data/kf/mimic_demo
   model:
     values:
       - ElasticNet
       - LGBMRegressor
       - GRU
       - LSTM
       - TCN
-      - Transformer
+      - TFT
```
Comment (rvandewater) [inline experiments/demo_benchmark_regression.yml:31]: Perhaps we can create a (or a couple) of yml files for the XAI experiments specifically?

File: icu_benchmarks/cross_validation.py [hunk 19]
```diff
@@ -37,7 +37,12 @@ def execute_repeated_cv(
     cpu: bool = False,
     verbose: bool = False,
     wandb: bool = False,
-    complete_train: bool = False
+    complete_train: bool = False,
```
Comment (rvandewater) [inline icu_benchmarks/cross_validation.py:40]: would it be possible to use inheritance here? I.e., we have a default CV and then one specifically for your implementation.

File: icu_benchmarks/models/utils.py [hunk 20]
```diff
@@ -188,3 +189,379 @@ def version(self):
     @rank_zero_only
     def log_hyperparams(self, params):
         pass
+
+
+def Faithfulness_Correlation(
```
Comment (rvandewater) [inline icu_benchmarks/models/utils.py:194]: Would clearly indicate that the code comes from Quantus

File: icu_benchmarks/run.py (hunk 6 above)
Comment (rvandewater) [reply icu_benchmarks/run.py:85]: Perhaps we can integrate this into a .gin file binding? So then we don't clutter up the run for people who do not necessarily want to use this.

File: icu_benchmarks/tuning/hyperparameters.py (hunk 9 above)
Comment (rvandewater) [reply icu_benchmarks/tuning/hyperparameters.py:39]: Let's see if we can change this for the "more integrated approach"

File: experiments/demo_benchmark_classification.yml (hunk 16 above)
Comment (youssefmecky96) [reply experiments/demo_benchmark_classification.yml:22]: I think this was before I edited any code but I can look it in for sure

File: experiments/demo_benchmark_regression.yml (hunk 18 above)
Comment (youssefmecky96) [reply experiments/demo_benchmark_regression.yml:31]: I think that would make it cleaner for sure once we have the setup for the models

File: environment.yml (hunk 17 above)
Comment (youssefmecky96) [reply environment.yml:16]: I think this was before I edited any code but I can look it in for sure same as the comment before

File: icu_benchmarks/cross_validation.py (hunk 19 above)
Comment (youssefmecky96) [reply icu_benchmarks/cross_validation.py:40]: Would work I think yes

File: icu_benchmarks/models/utils.py (hunk 20 above)
Comment (youssefmecky96) [reply icu_benchmarks/models/utils.py:194]: I already reference quantus in the function however the code here is adapted from the quantus implementation to get it work with captum and the model

File: icu_benchmarks/run_utils.py [hunk 21]
```diff
@@ -52,6 +53,14 @@ def build_parser() -> ArgumentParser:
     parser.add_argument("-sn", "--source-name", type=Path, help="Name of the source dataset.")
     parser.add_argument("--source-dir", type=Path, help="Directory containing gin and model weights.")
     parser.add_argument("-sa", "--samples", type=int, default=None, help="Number of samples to use for evaluation.")
+    parser.add_argument("--explain", default=False, action=BOA, help="Provide explaintations for predictions.")
```
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply icu_benchmarks/run_utils.py:56]: will fix these

File: scripts/plotting/plotting.py [hunk 22]
```diff
@@ -49,3 +50,194 @@ def calibration_curve(self):
...
+                marker="o",
+                color="skyblue",
+                linestyle="-",
+                linewidth=2,
+                markersize=8,
+            )
+            plt.xlabel("Feature")
+            plt.ylabel("{} Attribution".format(method_name))
+            plt.title("{} Attribution Values".format(method_name))
+            plt.xticks(
```
[the remark this answers is not in the document]
Comment (youssefmecky96) [reply scripts/plotting/plotting.py:161]: for sure can be the next steps for plotting
